netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -128,29 +133,52 @@ public AsciiString scheme() {
     };
   }
 
+  public static ProtocolNegotiator serverTls(final SslContext sslContext) {
+    return serverTls(sslContext, null);
+  }
+
   static final class ServerTlsHandler extends ChannelInboundHandlerAdapter {
     private final ChannelHandler next;
     private final SslContext sslContext;
+    private ObjectPool<? extends Executor> executorPool;
+    private Executor executor;
 
     private ProtocolNegotiationEvent pne = ProtocolNegotiationEvent.DEFAULT;
 
-    ServerTlsHandler(ChannelHandler next, SslContext sslContext) {
+    ServerTlsHandler(ChannelHandler next, SslContext sslContext,
+        ObjectPool<? extends Executor> executorPool) {
       this.sslContext = checkNotNull(sslContext, "sslContext");
       this.next = checkNotNull(next, "next");
+      this.executorPool = executorPool;
+    }
+
+    ServerTlsHandler(ChannelHandler next, SslContext sslContext) {
+      this(next, sslContext, null);
     }
 
     @Override
     public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
       super.handlerAdded(ctx);
       SSLEngine sslEngine = sslContext.newEngine(ctx.alloc());
-      ctx.pipeline().addBefore(ctx.name(), null, new SslHandler(sslEngine, false));
+      if (this.executorPool != null
+          && SystemPropertyUtil.getBoolean("io.netty.handler.ssl.openssl.useTasks", false)
+          == true) {
+        this.executor = this.executorPool.getObject();
+        ctx.pipeline().addBefore(ctx.name(), null, new SslHandler(sslEngine, false,
+            this.executor));
+      } else {
+        ctx.pipeline().addBefore(ctx.name(), null, new SslHandler(sslEngine, false));
+      }
     }
 
     @Override
     public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
       if (evt instanceof ProtocolNegotiationEvent) {
         pne = (ProtocolNegotiationEvent) evt;
       } else if (evt instanceof SslHandshakeCompletionEvent) {
+        if (this.executor != null && this.executorPool != null) {
+          this.executorPool.returnObject(this.executor);
```

Comment:
So the `executor` is returned to the pool only if `SslHandshakeCompletionEvent` is received (success or failure). Is it possible for the `ServerTlsHandler` (or `ClientTlsHandler`) to be removed without ever receiving this event? In that case there is a resource leak?

Reply:
Thanks for the comments Sanjay! 
Yeah my previous assumption was SslHandshakeCompletionEvent will always be passed through and if that's not the case, we might end up with a resource leak. A better way might be to make use of the close() function in ClientTlsProtocolNegotiator and the equivalent anonymous part in serverTls. I will update the code accordingly, and add some tests.

Reply:
Reading the Javadoc for `ProtocolNegotiator.close` : "Called when the Channel transitions to terminated. Is currently only supported on client-side; server-side protocol negotiators will not see this method called."

So this gets called only on channel termination and not for server-side protocol negotiator.

Reply:
Yeah I think this needs to be further discussed before I could move on.
@ejona86 I remembered we talked about this in our last meeting and you suggested putting the cleaning up functions into the close() method. But since it is not guaranteed on server side that this function will be called when TLS handshake ends, could you please provide more insights on how to handle this for server side? 
Any suggestions would be really appreciated :-)

Reply:
I would suggest putting any cleanup code in handlerRemoved().